### PR TITLE
7176515: ExceptionInInitializerError for an enum with multiple switch statements

### PR DESCRIPTION
**Pull request by archiecobbs**
This patch is both a minor optimization and a fix for JDK-7176515.

JDK-7176515 relates to how javac handles `switch` statements on `Enum` values, which are effectively "syntactic sugar".

The simple approach would be to just get the enum's `ordinal()` value and then revert to a normal integer value switch on that. However, this snapshots the ordinal values at compile-time, and thus can fail if the `Enum` class is recompiled later.

Presumably to avoid that possibility (are there any other reasons?) the compiler instead uses a lookup table. This table is dynamically constructed at runtime by a static initializer in a synthetic class. This avoids the "stale ordinal" problem of the simple approach, but it also creates a potential class initialization loop if there happens to an `Enum` switch inside an `Enum` class constructor, as demonstrated by the bug.

This patch makes the following change: If we are handling an `Enum` switch, and the `Enum` class we are switching on is also the class we are compiling, then just go with the simple approach, because the "stale ordinal" problem can't happen in this case so there's no need to build a runtime lookup table.

This results in two improvements:
* It avoids building the lookup table for switches on self
* It fixes JDK-7176515 as stated

Although the reproducing test case provided in JDK-7176515 gets fixed by this patch, the underlying issue is still there and can still be triggered with a slightly more complicated test case (included, but commented out).

So JDK-7176515 could be left open (or a new bug created) and a separate discussion had about how to "really" fix it.

Part of this discussion should be defining what that means, i.e., the boundaries of the bug. There are some scenarios that are basically impossible to fix, for example, two `Enum` classes whose constructors take as a parameter instances of the other `Enum` class and then switch on it. That cycle has nothing to do with how switches are handled.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-7176515](https://bugs.openjdk.org/browse/JDK-7176515): ExceptionInInitializerError for an enum with multiple switch statements
 * [JDK-8299760](https://bugs.openjdk.org/browse/JDK-8299760): ExceptionInInitializerError for an enum with multiple switch statements, follow-up

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [05fb5e76](https://git.openjdk.org/jdk/pull/10797/files/05fb5e76f48ec314e1bb1982a34031ee62ee3c9c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/10797/head:pull/10797` \
`$ git checkout pull/10797`

Update a local copy of the PR: \
`$ git checkout pull/10797` \
`$ git pull https://git.openjdk.org/jdk.git pull/10797/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10797`

View PR using the GUI difftool: \
`$ git pr show -t 10797`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10797.diff">https://git.openjdk.org/jdk/pull/10797.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/enum/T7176515.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+/*
+ * @test
+ * @bug 7176515
+ * @summary ExceptionInInitializerError for an enum with multiple switch statements
+ */
+
+import java.math.RoundingMode;
+
+public class T7176515 {
```
Comment (jddarcy) [inline test/langtools/tools/javac/enum/T7176515.java:32]: We've generally moved away from naming test after the bug number, although that was a long-standing practice in the past.

File: test/langtools/tools/javac/enum/T7176515.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+//            default:
+//                break;
+//            }
+        }
+    }
+
+    public static void main(String[] args) {
+        boolean shouldBeOdd = true;
+        for (MyEnum x : MyEnum.values()) {
+            assert x.isOdd() == shouldBeOdd;
```
Comment (jddarcy) [inline test/langtools/tools/javac/enum/T7176515.java:90]: While we do typically run the tests with assertions enabled, it is preferable if the tests did not rely on that for proper operation.

File: test/langtools/tools/javac/enum/T7176515.java (hunk 1 above)
Comment (archiecobbs) [reply test/langtools/tools/javac/enum/T7176515.java]: Thanks for the heads-up. Test name updated in 8cd5b49c5f2.

File: test/langtools/tools/javac/enum/T7176515.java (hunk 2 above)
Comment (archiecobbs) [reply test/langtools/tools/javac/enum/T7176515.java]: Good idea - thanks. Fixed in 8cd5b49c5f2.

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Comment by vicente-romero-oracle** [comment]
I think that a follow-up JIRA issue should be filed to track the situation in which the bug can still occur

**Comment by vicente-romero-oracle** [comment]
> > I think that a follow-up JIRA issue should be filed to track the situation in which the bug can still occur
> 
> Agreed (though it's not something I have permission to do).

true, I have created: https://bugs.openjdk.org/browse/JDK-8299760

**Comment by archiecobbs** [comment]
I've made a couple of improvements to this patch. As a result the follow-up bug [8299760](https://bugs.openjdk.org/browse/JDK-8299760), and also [8219412 - Eager enum class initialization with enum switch](https://bugs.openjdk.org/browse/JDK-8219412) are now fixed.

Improvement 1

The previous patch avoided generating a mapping table when the `enum` being switched on was the same class as the class currently being compiled. This is because there is no possibility of the ordinal values changing at runtime due to a recompilation. But this is too conservative - the same thing is true for any `enum` type defined in the same top-level class. So we generalize the test from "same class" as "in the same top-level class compilation". This change fixes the follow-up bug [8299760](https://bugs.openjdk.org/browse/JDK-8299760).

Improvement 2

All of the mapping tables for any `enum` type switched were being jammed into the same compiler-generated class. This meant that all of these `enum` types were initialized at the same time, regardless of which one actually experiences a 'first use'. This is what causes [8219412](https://bugs.openjdk.org/browse/JDK-8219412). The fix here is simple - put each mapping table into its own compiler-generated class.

/solves 8299760
/solves 8219412

**Comment by archiecobbs** [comment]
On second thought, the 2nd improvement belongs in a separate PR, as there probably is more discussion required.

Reverting that part of the change from this PR.

/issue remove 8219412

**Comment by vicente-romero-oracle** [comment]
I'm seeing one test failing with the last version of this patch:

- test/hotspot/jtreg/runtime/cds/appcds/jvmti/ClassFileLoadHookTest.java

**Comment by vicente-romero-oracle** [comment]
I'm still seeing test: test/hotspot/jtreg/runtime/cds/appcds/jvmti/ClassFileLoadHookTest.java failing with this patch

**Comment by archiecobbs** [comment]
> I'm still seeing test: test/hotspot/jtreg/runtime/cds/appcds/jvmti/ClassFileLoadHookTest.java failing with this patch

Strange... on github everything is passing now: https://github.com/archiecobbs/jdk/actions/runs/4473109724

But when I run that specific test on my laptop, it fails for me too.

So I guess this test is not included in the standard test suite?

In any case should be fixed in e6ef2acf8ae.

**Comment by jaikiran** [comment]
Hello Archie,

> > I'm still seeing test: test/hotspot/jtreg/runtime/cds/appcds/jvmti/ClassFileLoadHookTest.java failing with this patch
> 
> Strange... on github everything is passing now: https://github.com/archiecobbs/jdk/actions/runs/4473109724
> 
> But when I run that specific test on my laptop, it fails for me too.

GitHub actions job for the jdk repo is configured to run only `tier1` tests. The tests are grouped into different tiers in a file called `TEST.groups`. For hotspot that file resides here and the tier1 definition is this https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/TEST.groups#L545. You will notice that it's configured to run (among other things), the `tier1_runtime` which is defined in that same file https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/TEST.groups#L364. That group is defined to run everything under `runtime/` directory except the ones marked with a `-` (minus) sign and one of those is https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/TEST.groups#L404 `tier1_runtime_appcds_exclude` which is defined here https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/TEST.groups#L513, which except for a few tests, excludes everything else under `runtime/cds/appcds/` directory, so I believe that's why this `test/hotspot/jtreg/runtime/cds/appcds/jvmti/ClassFileLoadHookTest.java` isn't run in that job.

**Comment by archiecobbs** [comment]
> > > I'm still seeing test: test/hotspot/jtreg/runtime/cds/appcds/jvmti/ClassFileLoadHookTest.java failing with this patch
> > 
> > Strange... on github everything is passing now: https://github.com/archiecobbs/jdk/actions/runs/4473109724
> > But when I run that specific test on my laptop, it fails for me too.
> 
> GitHub actions job for the jdk repo is configured to run only `tier1` tests.

Thanks very much for that clarification.

**Pull request closed without merging**